### PR TITLE
[AIRFLOW-5542] Added BQ query results export to GCS  operator

### PR DESCRIPTION
**Pull request by meftasadat**
Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following AIRFLOW-5542 (https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. 
  - https://issues.apache.org/jira/browse/AIRFLOW-5522

### Description

- Executes a BigQuery Standard SQL query and exports query results to GCS.

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

File: airflow/contrib/operators/bigquery_export_query_results_to_gcs.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            cursor.run_query(destination_dataset_table=self.temp_table_name,
+                             write_disposition='WRITE_TRUNCATE',
+                             allow_large_results=True,
+                             sql=self.query,
+                             use_legacy_sql=False)
+            query_execution_success = True
+            cursor.run_extract(source_project_dataset_table=self.temp_table_name,
+                               destination_cloud_storage_uris=self.destination_cloud_storage_uris,
+                               compression=self.compression,
+                               export_format='CSV',
```
Comment (eladkal) [inline airflow/contrib/operators/bigquery_export_query_results_to_gcs.py]: Why CSV?
BigQuery supports Arrays / Record types. These can not be converted to CSV.

File: airflow/contrib/operators/bigquery_export_query_results_to_gcs.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    :param query: The BigQuery SQL to execute.
+    :type query: string
+    :param destination_cloud_storage_uris: The destination Google Cloud
+        Storage URI (e.g. gs://some-bucket/some-file.txt). (templated) Follows
+        convention defined here:
+        https://cloud.google.com/bigquery/exporting-data-from-bigquery#exportingmultiple
+    :type destination_cloud_storage_uris: List[str]
+    :param compression: Type of compression to use.
+    :type compression: str
+    :param export_format: File format to export.
```
Comment (eladkal) [inline airflow/contrib/operators/bigquery_export_query_results_to_gcs.py:47]: You take export_format as parameter but it's never being used.
You hard codded CSV later as the only option

File: airflow/contrib/operators/bigquery_export_query_results_to_gcs.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    :type labels: dict
+    :param location: The location used for the operation.
+    :type location: str
+    """
+    template_fields = ('query',
+                       'destination_cloud_storage_uris',
+                       'field_delimiter',
+                       'project_id',
+                       'bigquery_conn_id',
+                       'compression',
```
Comment (eladkal) [inline airflow/contrib/operators/bigquery_export_query_results_to_gcs.py]: compression templated?

File: airflow/contrib/operators/bigquery_export_query_results_to_gcs.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+                                        service=service)
+            cursor.create_empty_dataset(dataset_id=self.dataset_id,
+                                        project_id=self.project_id,
+                                        dataset_reference={'defaultTableExpirationMs': self.default_table_expiry_in_ms})
+            dataset_creation_success = True
+            cursor.run_query(destination_dataset_table=self.temp_table_name,
+                             write_disposition='WRITE_TRUNCATE',
+                             allow_large_results=True,
+                             sql=self.query,
+                             use_legacy_sql=False)
```
Comment (eladkal) [inline airflow/contrib/operators/bigquery_export_query_results_to_gcs.py:133]: If this is a temp table why not set a fail safe by defining expiration ? There can be a case where table is generated but not deleted due to error in deletion, isn't it?

File: airflow/contrib/operators/bigquery_export_query_results_to_gcs.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+class BigQueryExportQueryResultsToGCS(BaseOperator):
+    """
+    Executes a BigQuery Standard SQL query and exports query results to GCS.
+    For large query results, a GCS * pattern must be present
+    Grant necessary Service Account permission to your output GCS location table
+
+    :param project_id: The GCP project where the query gets executed, temp table is stored and extracted from
+    :type project_id: string
+    :param query: The BigQuery SQL to execute.
+    :type query: string
```
Comment (eladkal) [inline airflow/contrib/operators/bigquery_export_query_results_to_gcs.py:39]: Can the query be given as a SQL file? Like https://github.com/apache/airflow/blob/master/airflow/gcp/operators/bigquery.py#L367

**Review by eladkal — CHANGES_REQUESTED** [review]
I'm not really sure I get it...
1- Why can't it be an addition to the existed BigQueryToCloudStorageOperator? Why the need for a new operator for this functionality?

2 - If a new operator is required then why not reuse functionality that we already have? It can be achieved by using existed operators : 
      a. BigQueryOperator Runs query and save output to table
      b. Export table to storage with BigQueryToCloudStorageOperator
      c. delete table with BigQueryTableDeleteOperator
    This is pretty much the logic you tried to implement here. By using existed operators which are tested there is less chance for bugs and less code to test and maintain.

File: airflow/contrib/operators/bigquery_export_query_results_to_gcs.py (hunk 1 above)
Comment (meftasadat) [reply airflow/contrib/operators/bigquery_export_query_results_to_gcs.py]: I fixed this

File: airflow/contrib/operators/bigquery_export_query_results_to_gcs.py (hunk 2 above)
Comment (meftasadat) [reply airflow/contrib/operators/bigquery_export_query_results_to_gcs.py:47]: this is fixed now

File: airflow/contrib/operators/bigquery_export_query_results_to_gcs.py (hunk 4 above)
Comment (meftasadat) [reply airflow/contrib/operators/bigquery_export_query_results_to_gcs.py:133]: I actually have defined a default table expiry using default_table_expiry_in_ms in LOC#90

**Comment by meftasadat** [comment]
> I'm not really sure I get it...
> 1- Why can't it be an addition to the existed BigQueryToCloudStorageOperator? Why the need for a new operator for this functionality?
> 
> 2 - If a new operator is required then why not reuse functionality that we already have? It can be achieved by using existed operators :
> a. BigQueryOperator Runs query and save output to table
> b. Export table to storage with BigQueryToCloudStorageOperator
> c. delete table with BigQueryTableDeleteOperator
> This is pretty much the logic you tried to implement here. By using existed operators which are tested there is less chance for bugs and less code to test and maintain.

Thanks for reviewing this. I completely agree with you but the main motivation for the operator is to have less client-side code.

**Comment by kaxil** [comment]
I agree with the question raised. Don't see any added value of a new operator over here when same can be achieved by BigqueryToGCS operator

**Pull request closed without merging**